### PR TITLE
`lightning-net` crate

### PR DESCRIPTION
**Pull request by MaxFangX**
# lightning-net

## Motivation

@philip9 and I are working on an initiative to compile and run LDK within an SGX enclave using the Fortanix Enclave Development Platform (EDP). Among the many additional constraints imposed when trying to run Rust code inside SGX is that we cannot use Tokio, specifically none of the `net` features that are relied upon in the current Tokio-based implementation of the rust-lightning network stack `lightning-net-tokio` because those features rely on `mio` which in turn relies on `epoll`-like syscalls which are not available in the `x86_64-fortanix-unknown-sgx` (EDP) compilation target.

## Description

This PR implements a new crate called `lightning-net` which essentially reimplements the functionality of `lightning-net-tokio` but without using Tokio, futures, async Rust, or [anything else](https://edp.fortanix.com/docs/concepts/rust-std/) that will not compile to the `x86_64-fortanix-unknown-sgx` target.

As the code currently stands, the ldk-sample node could easily swap out all usages of `lightning-net-tokio` with their equivalent functions in `lightning-net` and function perfectly fine. I've tested this already in a private fork of ldk-sample running on testnet - let me know if you'd like me to make a public version so that you can play around with it as well.

### Previous work

Prior to this pull request, @phlip9 and I iterated on this module on our own fork at [lexe-tech/rust-lightning#1](https://github.com/lexe-tech/rust-lightning/pull/1). We now feel it is ready for the next stage of review prior to being upstreamed.

### Follow-on work

Currently, `lightning-net` does not yet implement any SGX-specific functionality. In particular, we need to implement an SGX-compatible socket that has a subset of the existing `TcpStream` features. Follow-on work will implement this concrete `SgxTcpStream` using either [custom Fortanix Usercalls](https://edp.fortanix.com/docs/api/fortanix_sgx_abi/struct.Usercalls.html) (or possibly via patches to Rust `std`), and modify the `handle_connection()` and `initiate_outbound()` functions to take in an impl of e.g. a new `Socket` trait. Since we are not yet settled on the final / best approach, and since this pull request is fairly beefy already at +1129 lines, we figure it would be best to upstream this code now and add in the generics / SGX functionality later.

## Original commits prior to squash

Squashed commits in original PR at [lexe-tech/rust-lightning#1](https://github.com/lexe-tech/rust-lightning/pull/1), reviewed by @phlip9:
- Create empty lightning-net module
- SyncSocketDescriptor skeleton implementation
- Connection struct with read and write halves
- Implement shutdown() for TcpReader and TcpWriter
- Implement most of spawn_inbound_handler
- Add crossbeam and some associated channels
- Implement basic functionality of ConnectionWriter
- Pass a Arc<PeerManager> into ConnectionReader
- Get a SyncSocketDescriptor into the ConnectionWriter
- Implement first half of pausing reads
- Renome write tx/rx to write_data tx/rx
- Rename ConnectionReader/Writer to Reader/Writer
- Create the Reader/Writer cmd channels
- Send ResumeRead command to the Reader
- Remove write_cmd, rename reader_cmd -> resume_read
- Reader can handle ResumeRead events
- Implement disconnecting from peer
- send_data() actually sends data now
- Allow send_data() to pause reads
- Get a Arc<PeerManager> into Writer
- Refactor write_data tx/rx into writer_cmd tx/rx
- Give Reader/Writer a cmd tx for the other
- Implement all disconnecting except TcpDisconnectooor
- Remove the Arc<Mutex<Connection>>
- Refactor Connection into setup()
- disconnect_socket() now calls into TcpDisconnectooor
- Get a SyncSocketDescriptor into Writer
- Call write_buffer_space_avail and socket_disconnected v2
- resume_read check should go before the early return
- Handle read() ErrorKind variants
- Implement handle_connection and initiate_outbound
- Finish writing doc comments, clean up
- Basic tests for lightning-net
- These is modeled exactly after the tests in lightning-net-tokio
- Update GitHub CI and dependabot for lightning-net
- Reduce the dependencies used
- Implementing PR review feedback from @phlip9
- Remove some comments about SGX
- Hide ID_COUNTER in a function, return JoinHandles
- Reader/Writer now send commands via the descriptor
- Extract do_recv(), do_try_recv(), do_read() fns v2
- Split WriteData command into its own channel v4
- Remove `std::net` limitations from EDP doc comment
- Implement ReaderState enum v2

### CODE REVIEW

**Comment by MaxFangX** [comment]
Looks like a bunch of stuff broke after rebasing, on it.

File: lightning-net/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1098 @@
...
+    /// Handles a `ReaderCommand`.
+    fn handle_command(&mut self, cmd: ReaderCommand) {
+        use ReaderCommand::*;
+        use ReaderState::*;
+
+        match (cmd, &self.state) {
+            (_, &ShuttingDown) => {}
+            (PauseRead, _) => self.state = Paused,
+            (ResumeRead, _) => self.state = Reading,
+            (Shutdown, _) => self.state = ShuttingDown,
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:519]: The `Shutdown` message was implemented because while reading is paused, we don't want the `Reader` to be, well, reading. The reader needs to block on _something_, otherwise there is no way for it to know to resume reading (or to shut down, for that matter) without a busy loop. So that's why the `reader_cmd` channel is necessary. But since the Reader can be blocked on the channel instead of `read()`, it would miss any `TcpStream::shutdown()`s initiated by the `Writer` or `PeerManager`. Thus, the `Shutdown` command which can unblock the `Reader` while it is stuck on `recv()`.

File: lightning-net/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1098 @@
...
+        // PeerManager rejected this connection; disconnect the TcpStream and
+        // don't even start the Reader and Writer.
+        descriptor.disconnect_socket();
+        // In line with the requirements of new_inbound_connection() and
+        // new_outbound_connection(), we do NOT call socket_disconnected() here.
+        e
+    })
+}
+
+// NOTE: It would be nice to have a `impl From<SocketAddr> for NetAddress` in
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:212]: Sweet! Will replace the equivalent functionality in `lightning-net-tokio` as well.

File: lightning-net/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:519]: The doc comment for `SyncSocketDescriptor::disconnect_socket()` explains why the current design uses channels:

(The below is after fixing a typo)
```
    /// Shuts down the Reader, Writer, and the underlying TcpStream.
    ///
    /// There are several ways that a disconnect might be triggered:
    /// 1) The Reader receives Ok(0) from TcpStream::read() (i.e. the
    ///    peer disconnected), or an Err(io::Error) that shouldn't be retried.
    /// 2) The Reader receives Err from PeerManager::read_event(); i.e.
    ///    Rust-Lightning told us to disconnect from the peer.
    /// 3) The Writer receives Ok(0) from TcpStream::write() (undocumented
    ///    behavior), or an Err(io::Error) that shouldn't be retried.
    /// 4) The Writer receives Err from PeerManager::write_buffer_space_avail();
    ///    Rust-Lightning told us to disconnect from the peer.
    /// 5) This function is called.
    ///
    /// The disconnect will be handled differently depending on the source of
    /// the trigger:
    /// - (1) and (2): If the Reader received the trigger, it will shut down
    ///   BOTH halves of the shared TcpStream AND send a Shutdown command to the
    ///   Writer.
    ///
    ///   - The explicit Shutdown command from the Reader is necessary because
    ///     if the Writer is blocked on `writer_cmd_rx.recv()` due to its
    ///     internal buffer being empty, the only way it can be unblocked is by
    ///     receiving a command, in this case the Shutdown command.
    ///   - The Reader closing both halves of the TCP stream is necessary
    ///     because while the writer is blocked on write(), the only way it can
    ///     unblock is by detecting the TCP disconnect.
    ///
    /// - (3) and (4): If the Writer received the trigger, it will shut down
    ///   BOTH halves of the shared TcpStream AND send a Shutdown command to the
    ///   Reader.
    ///
    ///   - The explicit Shutdown command from the Writer is necessary because
    ///     if the Reader is blocked on `reader_cmd_rx.recv()` due to
    ///     `read_paused == true`, the only way it can be unblocked is by
    ///     receiving a command, in this case the Shutdown command.
    ///   - The Writer closing both halves of the TCP stream is necessary
    ///     because while the reader is blocked on read(), the only way it can
    ///     unblock is by detecting the TCP disconnect.
    ///
    /// - (5): If the disconnect was initiated here, a Shutdown command will be
    ///   sent to both the Reader and the Writer, AND the TcpDisconnectooor will
    ///   shutdown both halves of the shared TCP stream. The Shutdown command
    ///   ensures that the Reader / Writer will unblock if they are currently
    ///   blocked on `recv()`. The TCP stream shutdown ensures that they will
    ///   unblock if they are currently blocked on `read()` or `write()`
    ///   respectively.
    ///
    /// In cases (1) and (3), the disconnect was NOT initiated by
    /// Rust-Lightning, so the Reader / Writer notify the PeerManager using
    /// `socket_disconnected()`.
```

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:519]: Re `TcpStream::shutdown()` noop in the SGX docs - nice attention to detail, you actually read the docs! Indeed, the code in its current state will not function correctly inside SGX. The `Socket` trait that we plan to introduce in the will will have a `shutdown()` method. That way, we can swap out the `std::net::TcpStream` with our own `SgxTcpStream` which uses an EDP usercall to implement `shutdown()`.

The remaining comments about `std::time` and `std::thread` are in the `lightning-net` doc comments so that future implementers know to avoid using those in the rest of the crate, since the `SgxTcpStream` will still rely on this code.

File: lightning-net/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1098 @@
...
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+
+use core::hash;
+use std::io::{Read, Write};
+use std::net::{IpAddr, Shutdown, SocketAddr, TcpStream};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::thread::{self, JoinHandle};
+
+use crossbeam_channel::{select, Receiver, Sender, TryRecvError, TrySendError};
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:54]: We could use the `std` channels, but the other thing used from `crossbeam-channel` is the `select!` macro which operates on crossbeam channel `recv()`s rather than futures. There does not appear to be a `select!` macro (or equivalent functionality) available from `std`.

One approach could be to remove the `select!` by sending `WriteData` requests and `WriterCommand::Shutdown` commands in the same channel, but this introduces a possible race condition due to the need to maintain a bounded `WriteData` queue. What happens is that instead of having a bounded channel of size 1 that only holds `WriteData` requests, we need to have it be of size 2 so that we can always leave a space open for a `Shutdown` command. But in order to leave that second space open, `send_data()` first needs to check `writer_cmd_tx.is_empty()` before actually pushing a `WriteData<Vec<u8>>` into the channel. This could create a TOCTTOU race condition where something else could push into the `writer_cmd` channel in between the `is_empty()` check and the `try_send()` that pushes the write request. Which in turn means that there might not be space for a `Shutdown` command to enter the channel when pushed, which means that a `Reader` or `Writer` thread might be left hanging.

Having only one channel to the Writer was how I implemented the writer commands previously, before splitting it out into two channels to remove the race condition and reduce complexity. [Here's the commit](https://github.com/lexe-tech/rust-lightning/pull/1/commits/6462dbc54a788969bc69141b492d88202dfa4ecb) where I did that, and [here's](https://github.com/lexe-tech/rust-lightning/pull/1/commits/6462dbc54a788969bc69141b492d88202dfa4ecb#diff-6062c086b39587c4e593eeec4310374348ec5082cf9f3bc456ce1f4da53b834fL359) the `is_empty()` check I was referring to.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:54]: Hmm, I suppose the `writer_cmd` channel capacity could still just be increased to 3. I think the current design is much cleaner and simpler (crossbeam is also more performant), but I'll let you decide.

File: lightning-net/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1098 @@
...
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    // Init channels
+    let (reader_cmd_tx, reader_cmd_rx, writer_cmd_tx, writer_cmd_rx, write_data_tx, write_data_rx) =
+        init_channels();
+
+    // Generate a new ID that represents this connection
+    let conn_id = next_connection_id();
+    let socket_addr = stream.peer_addr().unwrap();
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs]: Done

File: lightning-net/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,1098 @@
...
+    peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+    descriptor: SyncSocketDescriptor,
+    writer_cmd_rx: Receiver<WriterCommand>,
+    write_data_rx: Receiver<Vec<u8>>,
+    /// An internal buffer which stores the data that the Writer is
+    /// currently attempting to write.
+    ///
+    /// This buffer is necessary because calls to self.inner.write() may fail or
+    /// may write only part of the data.
+    buf: Option<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: > I think you should be able to do the writer with no separate thread at all, either!

If I'm understanding you correctly, you mean to use `TcpStream`'s `set_nonblocking(true)`, and then do `stream.write()` directly within `SyncSocketDescriptor::send_data()`? That would work for the writer, but since the `Reader` and `Writer` share the same underlying `TcpStream` which is managed by the OS, this means that the reader would no longer be able to block on `read()`, and could only read in a busy loop. From the [`set_nonblocking()` docs](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_nonblocking):

> This will result in `read`, `write`, `recv` and `send` operations becoming nonblocking, i.e., immediately returning from their calls. If the IO operation is successful, `Ok` is returned and no further action is required. If the IO operation could not be completed and needs to be retried, an error with kind [`io::ErrorKind::WouldBlock`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WouldBlock) is returned.
> 
> On Unix platforms, calling this method corresponds to calling `fcntl FIONBIO`. On Windows calling this method corresponds to calling `ioctlsocket FIONBIO`

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: > rust-lightning should re-call with the remaining data if you only write part

This requires `send_data()` to know immediately whether a call to `write()` wrote all of the bytes, some of the bytes, etc. But since we cannot use non-blocking writes due to the constraints on the `Reader` above, there needs to be a dedicated `Writer` thread that can block on `write()`. Which AFAICT necessitates a layer of indirection (i.e. the `write_data` channel) between `send_data()` and the `Writer`, which allows the `Writer` to receive write requests while it is blocked on `write(), but which unfortunately makes partial `write()`s undetectable to callers of `send_data()`.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: > I think you should be able to do the writer with no internal buffer at all

I could imagine making the `write_data` channel a zero-capacity channel, so that when the `Writer` `recv()`s data, it immediately turns around and passes the data into `write()`. But since `send_data()` cannot detect partial writes (as explored above), the `Writer` must continue trying to send the remaining data until it succeeds (or shuts down), since the call to `send_data()` reported to Rust-Lightning that the entire write was a success. And in order to do that, the remaining data must be stored in memory somewhere, thus the internal buffer.

I don't see a way around these constraints - any suggestions? The constraints are subtle but have wide-reaching implications regarding the design of this crate.

File: lightning-net/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (phlip9) [reply lightning-net/src/lib.rs:54]: My understanding is that the [`std::mpsc` channels are unmaintained and buggy](https://github.com/rust-lang/rust/issues/39364) : (

My hope is that https://github.com/rust-lang/rust/pull/93563 will go through and just replace the current `std::mpsc` with `crossbeam-channel`. In the meantime, it feels safer to just depend on `crossbeam-channel`.

File: lightning-net/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,1098 @@
...
+        let _ = self
+            .descriptor
+            .writer_cmd_tx
+            .try_send(WriterCommand::Shutdown);
+    }
+
+    /// Checks for a command in a non-blocking manner, handling the command
+    /// if there was one.
+    fn do_try_recv(&mut self) {
+        match self.reader_cmd_rx.try_recv() {
```
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:478]: I tried implementing this, but ran into the issue that when reading is paused, the `Reader` _needs_ something else to block on, otherwise it requires a busy loop on `should_read.load()` in order to detect that it should resume reads.

**Comment by MaxFangX** [comment]
After the review by @TheBlueMatt, I think this module needs more comprehensive tests. A future contributor is likely not to be as familiar with all the constraints around threads blocking on `recv`, `read`, and `write`, and may easily introduce a bug that prevents a reader or writer thread from shutting down while reads are paused, the writer is writing, etc. I'll work on covering those cases next week.

File: lightning-net/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: If we set a timeout and `write()` times out, there's no way for the `PeerManager` to know that there is `write_buffer_space_avail()`. After the `Writer` completes a `write()`, it `recv()`s the next `Vec<u8>`, at which point it can notify the `PeerManager`. But without futures or polling, a failed `send_data()` call does not have a mechanism for making that call. Yet another wrinkle of using 100% synchronous code...

Also, timeouts [aren't configured at all](https://edp.fortanix.com/docs/concepts/rust-std/) when compiling to the EDP target.

> The following functions don't configure a timeout at all and revert back to default behavior:
>
> `net::TcpStream::connect_timeout`
> `net::TcpStream::set_read_timeout`
> `net::TcpStream::set_write_timeout`

This would mean that the future `Socket` trait would implementing these functions too. Right now, we are using only the most basic and common POSIX APIs - `connect`, `read`, `write`, `shutdown`. The more advanced functionality we use, the more functions we need to include into the `Socket` trait, which in turn means more burden on users that need to make the `lightning-net` work on their specific platform, possibly highly constrained platform which doesn't have access to as rich of an API set from the OS.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: For example, [this POSIX socket](https://docs.rs/posix-socket/latest/posix_socket/struct.Socket.html) does not support `connect_timeout`.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: The `Writer` channels could be replaced with something like `Arc<(Mutex<(Option<Vec<u8>>, bool)>, Condvar)>` (this is my first exposure to condition variables) but I don't see how we can get rid of the `Writer` thread.

File: lightning-net/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:478]: On it. Starting to think that we can indeed get rid of the entirety of the `crossbeam` dependency

File: lightning-net/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: Refreshed and only just saw your last four messages

> This is kinda nonsense, though - the os can always "deny service to an enclave" but just.....not context-switching to that process? Like, what does TCP timeouts have to do with it...

Yeah, it's a pain... I don't know what Fortanix's exact rationale is, but as the ecosystem currently stands the alternatives are to (1) make `Tokio` run inside SGX by fixing the entire Tokio stack down to `mio` or (2) to patch Rust `std`.

> Two threads per connection is a lot of overhead.

FWIW, for our use case it's not a problem at all, since our users only really need to connect to an LSP. I figure that if someone truly needs performance, they would be using Tokio and therefore `lightning-net-tokio`. This crate can serve people who need LDK to run on uncommon or not-well-supported platforms. And the way to accomplish that IMO is to keep the API to the OS extremely simple.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: I'll take a stab at replacing all of the `crossbeam` channels with shared state + condition variables, which should also eliminate the `select!`, and therefore the entire dependency. Beyond that, the only thing left to make this run inside SGX is to implement the custom usercalls. So I don't think there's that much more work to do actually.

[the remark this answers is not in the document]
Comment (MaxFangX) [reply lightning-net/src/lib.rs:574]: > To rethink this somewhat further, instead of jumping through hoops here, we could also consider running the actual net logic outside of the enclave - it doesn't change the security model to make the syscalls outside of the enclave, we just pass the stream of read/write_buffer_space_avail events and write calls over some IPC. I'd bet we could do all that with less total LoC :/

That's essentially what we will be doing once the usercalls are implemented: making the syscalls outside of the enclave. Maybe take a quick look at [the EDP architecture](https://edp.fortanix.com/docs/concepts/architecture/) (it's short)

**Comment by MaxFangX** [comment]
Please excuse the delay; I got my computer repaired during the last week while I was with family.

I still don't think the writer thread can be eliminated since there would be no way to call `write_buffer_space_avail`, but I did manage to replace all the crossbeam channels with shared state + condition variables. If the changes look good to you I can start writing some more comprehensive tests and docs @TheBlueMatt.

**Comment by MaxFangX** [comment]
> Right, I'm still somewhat curious as to the exact SGX programming environment here - are you able to define your own "usercall"s, and if so, can we define a select usercall? Alternatively, does it make more sense to run the network stack directly in normal user mode, with the in-SGX just handling a stream of read/write/writeable events? In either model we could run all the networking stuff for an arbitrary number of connections with one or two threads and then have the SGX stuff just pass the network bytes through to the PeerManager.

The custom usercalls are fairly limited. The main interface is the `UsercallExtension` trait defined [here](https://edp.fortanix.com/docs/api/enclave_runner/usercalls/trait.UsercallExtension.html) which exposes `connect_stream` and `bind_stream` which one can use to override the normal behavior of `TcpStream::connect()`.

Essentially, I'd have to implement an RPC service from scratch. The enclave would call `UsercallExtension::connect_stream("rpc", ...)` to obtain a client that can send bytes to the RPC service running on the host. The client would need to serialize all Read / Write events to bytes before sending it to the RPC service, and the RPC service would need to deserialize again before passing the actual Lightning data bytes to the peer address specified in the Read / Write event. The RPC service runs outside of the enclave and can use Tokio, so there would only need to be a one or two threads for handling all of the connections.

I think this is totally unnecessary though. On my lowest tier SGX Azure instance, the default maximum number of threads supported by the operating system is 31368. With `bitcoind` running and nothing else, there are 162 threads available. Assuming each lightning node requires 1 main thread and 2 threads per peer, this single instance can support 10,000 nodes running *simultaneously*, or it can support a single node with 15,000 P2P connections. Since the nodes really don't need to be running in the enclave all the time - they only need to be loaded in when receiving payments - the scenarios where a single cloud instance is running 10,000 nodes **simultaneously** or a single node is maintaining 15,000 P2P connections sounds absurd to me, even for the beefiest of node operators. Two threads per P2P connection is not a lot of overhead on modern infrastructures.

The quickest path forward

In short, yes, I think it's possible

**Comment by MaxFangX** [comment]
> Right, I'm still somewhat curious as to the exact SGX programming environment here - are you able to define your own "usercall"s, and if so, can we define a select usercall? Alternatively, does it make more sense to run the network stack directly in normal user mode, with the in-SGX just handling a stream of read/write/writeable events? In either model we could run all the networking stuff for an arbitrary number of connections with one or two threads and then have the SGX stuff just pass the network bytes through to the PeerManager.

The custom usercalls are fairly limited. The main interface is the `UsercallExtension` trait defined [here](https://edp.fortanix.com/docs/api/enclave_runner/usercalls/trait.UsercallExtension.html) which exposes `connect_stream` and `bind_stream` which one can use to override the normal behavior of `TcpStream::connect()`.

Essentially, I'd have to implement an RPC service from scratch. The enclave would call `UsercallExtension::connect_stream("rpc", ...)` to obtain a client that can send bytes to the RPC service running on the host. The client would need to serialize all Read / Write events to bytes before sending it to the RPC service, and the RPC service would need to deserialize again before passing the actual Lightning data bytes to the peer address specified in the Read / Write event. The RPC service runs outside of the enclave and can use Tokio, so there would only need to be a one or two threads for handling all of the connections.

I think this is totally unnecessary though. On my lowest tier SGX Azure instance, the default maximum number of threads supported by the operating system is 31368. With `bitcoind` running and nothing else, there are 162 threads available. Assuming each lightning node requires 1 main thread and 2 threads per peer, this single instance can support 10,000 nodes running *simultaneously*, or it can support a single node with 15,000 P2P connections. Since the nodes really don't need to be running in the enclave all the time - they only need to be loaded in when receiving payments - the scenarios where a single cloud instance is running 10,000 nodes **simultaneously** or a single node is maintaining 15,000 P2P connections sounds absurd to me, even for the beefiest of node operators. Two threads per P2P connection is not a lot of overhead on modern infrastructures.

Making the existing implementation work has already stalled our project by a few weeks. The easiest, lowest resistance path forward I see is to simply move ahead with the current implementation that uses two threads per connection, and implement a usercall to handle `TcpStream::shutdown()`. I think anyone who needs to run 20,000 nodes simultaneously is perfectly capable of using `lightning-net-tokio`, or creating their own bespoke implementation of the network stack.

**Comment by MaxFangX** [comment]
https://doc.rust-lang.org/std/thread/struct.Builder.html#method.stack_size

Thoughts on setting the thread stack size to eg 4KB? The biggest local variable is the 8kb buffer for the Reader, which can be heap allocated by using a Vec<u8> instead.

**Comment by MaxFangX** [comment]
> I'd be willing to bet the maximum stack depth for calling LDK functions is at least 256K-512K, at least once you count OS thread setup (likely more in your case for SGX?) and syscalls. In general, though, I don't think there's a reason to reduce the stack size - the stack size is effectively a "maximum", which if we overrun it will cause (in the best case) a crash. Linux, however, can avoid actually allocating the stack space, or at least swap it out, until its used. Of course once its used you don't get the memory back.

Ah, I forgot the calls into PeerManager - good point.

> More generally, this also doesn't consider the OS overhead of threads, context switches, etc, which, sure, aren't much on modern machines, but its also not free. For running in SGX on a beefy machine its probably totally fine, but it may not be the direction we want to go for a general lightning-net crate.

Noted. I did some digging on a `select!` for TcpStream read / write, and ended up at rust-lang/rust#6842 (8 years old) which was superceded by rust-lang/rfcs/issues/1081 in 2016. The conclusion, essentially, was that the solution for non-blocking IO would be to use `async`/`await`, and to conduct async IO using crates external to rust `std`, such as `mio`. One reason for that is that Rust `std` needs to be maximally portable, and crates such as `mio` tend to have problems even on some of the more common platforms like Windows. Which comes back to why I ended up writing `lightning-net` in the first place - `lightning-net-tokio` was insufficient for my platform due to its dependence on `mio`, so I had to reimplement the network stack using only technologies that were available 8 years ago.

My point is - if this crate is to be upstreamed, I think the primary selling point should be that it is maximally portable with minimal reliance on any platform-specific APIs, rather than that it can match the performance of `lightning-net-tokio`. Maybe it can be renamed to `lightning-net-sync` (or similar) to reflect this posture, so that LDK users don't think that `lightning-net` is the primary, recommended crate instead of `lightning-net-tokio`. There could also be more notes on the performance implications of using this crate in the crate-level doc comments which discuss context switches, stack allocations, 2 threads per peer, etc. Additionally, after iterating on this together we've managed to produce a crate that has 0 dependencies outside of `std` (most importantly not using `tokio` which is massive), so this crate could also be useful for the especially security-conscious.

As for me and Philip's SGX project, we could certainly remove the overhead of dedicated threads by implementing a usercall extension that uses futures / polling / similar, but then the implementation would be specific to the SGX platform and it wouldn't make any sense to upstream it. The EDP people did mention the possibility of supporting `TcpStream::shutdown()` a few years ago in fortanix/rust-sgx#110 - presently I hope to push that forward and implement the usercall for `shutdown()` via a patch to Rust `std` rather than by introducing a `Socket` trait and `SgxTcpStream` in the `lightning-net-sync` crate. The result is that everything in `lightning-net-sync` is general purpose, with no mention of SGX or of a `Socket` trait that would make using the crate more unwieldy, and thus would be worthwhile to make available for anyone to use as a part of LDK.

**Comment by MaxFangX** [comment]
All understood, and good points. Thanks for being welcoming to new contributors and taking the time to iterate with me on this @TheBlueMatt; it was a pleasure, and I learned a lot :) Will go ahead and close this PR.
